generator/lib/src/field_element_ex.dart

```diff
@@ -137,27 +137,58 @@ extension FieldElementEx on FieldElement {
           );
         }
         */
+
+        if (type.realmType?.mapping.canBePrimaryKey != true) {
+          final file = span!.file;
+          final listOfValidTypes = RealmPropertyType.values //
+              .map((t) => t.mapping)
+              .where((m) => m.canBePrimaryKey)
+              .map((m) => m.type);
+
+          throw RealmInvalidGenerationSourceError(
+            'Realm only supports the @PrimaryKey annotation on fields of type\n${listOfValidTypes.join(', ')}\nas well as their nullable versions',
+            element: this,
+            primarySpan: typeSpan(file),
+            primaryLabel: "$modelTypeName is not a valid type here",
+            todo: "Change the type of '$displayName' or remove the @PrimaryKey annotation",
+          );
+        }
       }
 
-      // Validate indexes
-      if ((((primaryKey ?? indexed) != null) && !(type.realmType?.mapping.indexable ?? false)) || //
-          (primaryKey != null && (type.isDartCoreBool || type.isExactly<RealmValue>()))) {
+      final indexType = RealmIndexType.values.elementAt(indexed?.value.getField("indexType")!.getField("index")!.toIntValue() ?? RealmIndexType.none.index);
+
+      if (indexed != null) {
         final file = span!.file;
-        final annotation = (primaryKey ?? indexed)!.annotation;
-        final listOfValidTypes = RealmPropertyType.values //
-            .map((t) => t.mapping)
-            .where((m) => m.indexable && ((m.type != bool && m.type != RealmValue) || primaryKey == null))
-            .map((m) => m.type);
+        if (indexType == RealmIndexType.none) {
+          throw RealmInvalidGenerationSourceError("Don't use RealmIndexType.none for @Indexed annotations",
+              element: this,
+              primarySpan: typeSpan(file),
+              primaryLabel: '@Indexed(RealmIndexType.none) is not valid annotation here',
```

Comment:
It's possible if we introduce two types for that, but it seemed a bit of an overkill. The reason why I think it makes sense to have `.none` there is because the schema can also be introspected from the Realm file. So in case you're writing an app that can open a Realm file dynamically, we should be able to communicate the state of the indexes on each column.

Reply:
If I get it correctly, `@Indexed(RealmIndexType.none)` is equivalent to not having `@Indexed` annotation at all, right? Are we going to require  from our users to place this annotation on every field that is not indexed?

Reply:
No, and this PR does not imply that.

Reply:
@nielsenko is correct - `@Indexed(RealmIndexType.none)` is incorrect usage of the API which will result in a generator error. But the `.none` option on `RealmIndexType` exists so that you can do something like:

```dart
var property = realm.schema[0].properties[0];
if (property.indexType == RealmIndexType.none) {
  console.log("Property ${property.name} is not indexed");
}
```

I.e. if you're introspecting the schema, you may want to know the index status of the properties and `.none` is valid status then.

Reply:
I totally agree with @desistefanova here. 
`@Indexed(RealmIndexType.none)` is very weird API and we should disallow writing it. This should be a static restriction instead of runtime one. Btw I am really surprised @nielsenko is agreeing with this as a strong proponent of static compile time APIs. This one is really no brainer to change to a compile time one just remove the `RealmIndexType.none` from the enum.

As for the case @nirinchev mentions,  if one wants to inspect a property at runtime we will return `null` for its `index` flag. 
```dart
var property = realm.schema[0].properties[0]; 
if (property.indexType == null) { //or if (property.index == null) { or something similarly named.
console.log("Property ${property.name} is not indexed"); 
}
```

Reply:
I generally dislike exposing nullable enums as `null` can just be represented as a value within the enum. That being said, it's not a hill I'm willing to die on, so will change it.

Reply:
For the record I still am a strong proponent for static compile time APIs 😄 .. I usually don't use nullable enums either, preferring to have `.none` like value instead, but perhaps just having a `RealmIndexType? indexType` on `Property` is better in this case?

Wrt. to misuse, I just want to point out that the generator actually did catch it, if anyone used `RealmIndexType.none`.